File: pc01/pc01_07/c/pc01_07.c
/**
 * Check the check
 * pc110107
 * uva10196
 */

#include <stdio.h>

#define SIZE 8

int black_pawn_dirs[8][2] = {
	{-1, -1},
	{-1, 1},
	{0, 0}, {0, 0}, {0, 0}, {0, 0}, {0, 0}, {0, 0}
};

int white_pawn_dirs[8][2] = {
	{1, -1},
	{1, 1},
	{0, 0}, {0, 0}, {0, 0}, {0, 0}, {0, 0}, {0, 0}
};

int knight_dirs[8][2] = {
	{-2, -1}, {-2, 1},
	{2, -1}, {2, 1},
	{-1, -2}, {-1, 2},
	{1, -2}, {1, 2}
};

int rook_dirs[8][2] = {
	{-1, 0},
	{1, 0},
	{0, -1},
	{0, 1},
	{0, 0}, {0, 0}, {0, 0}, {0, 0}
};

int bishop_dirs[8][2] = {
	{-1, -1},
	{-1, 1},
	{1, -1},
	{1, 1},
	{0, 0}, {0, 0}, {0, 0}, {0, 0}
};

int queen_dirs[8][2] = {
	{-1, -1}, {-1, 0}, {-1, 1},
	{0, -1}, {0, 1},
	{1, -1}, {1, 0}, {1, 1}
};

typedef struct {
	int row;
	int col;
	char name;
	int directions[8][2];
	int repeat;
} piece_t;

typedef struct {
	char board[8][8];
	piece_t white_king;
	piece_t black_king;
} chess_t;

int is_valid_position(int row, int col) {
	return (row >= 0 && row < SIZE && col >= 0 && col < SIZE);
}

void init_piece(piece_t *piece, char name, int repeat, int directions[8][2]) {
	piece->name = name;
	piece->repeat = repeat;
	for (int i = 0; i < 8; i++) {
		piece->directions[i][0] = directions[i][0];
		piece->directions[i][1] = directions[i][1];
	}
}

int is_attacked_by_piece(chess_t chess, int is_white, piece_t piece) {
	piece_t king = is_white ? chess.white_king : chess.black_king;

	for (int i = 0; i < 8; i++) {
		int row_dir = piece.directions[i][0];
		int col_dir = piece.directions[i][1];

		if (row_dir == 0 && col_dir == 0) {
			continue;
		}

		int row = king.row;
		int col = king.col;

		if (piece.repeat) {
			row += row_dir;
			col += col_dir;
			while (is_valid_position(row, col)) {
				if (chess.board[row][col] != '.') {
					if (chess.board[row][col] == piece.name) {
						return 1;
					}
					break;
				}
				row += row_dir;
				col += col_dir;
			}
		} else {
			row += row_dir;
			col += col_dir;

			if (is_valid_position(row, col) && chess.board[row][col] == piece.name) {
				return 1;
			}
		}
	}
	return 0;
}

int is_in_check(chess_t chess, int is_white) {
	piece_t attacker;

	init_piece(&attacker, (is_white ? 'p' : 'P'), 0, (is_white ? black_pawn_dirs : white_pawn_dirs));
	if (is_attacked_by_piece(chess, is_white, attacker)) {
		return 1;
	}

	init_piece(&attacker, (is_white ? 'n' : 'N'), 0, knight_dirs);
	if (is_attacked_by_piece(chess, is_white, attacker)) {
		return 1;
	}

	init_piece(&attacker, (is_white ? 'r' : 'R'), 1, rook_dirs);
	if (is_attacked_by_piece(chess, is_white, attacker)) {
		return 1;
	}

	init_piece(&attacker, (is_white ? 'b' : 'B'), 1, bishop_dirs);
	if (is_attacked_by_piece(chess, is_white, attacker)) {
		return 1;
	}

	init_piece(&attacker, (is_white ? 'q' : 'Q'), 1, queen_dirs);
	if (is_attacked_by_piece(chess, is_white, attacker)) {
		return 1;
	}

	return 0;
}

int main(void) {
	char input_board[SIZE][SIZE];
	int game_count = 0;

	while (1) {
		int empty = 1;
		chess_t chess;

		for (int i = 0; i < SIZE; i++) {
			for (int j = 0; j < SIZE; j++) {
				scanf(" %c", &input_board[i][j]);
				chess.board[i][j] = input_board[i][j];

				if (input_board[i][j] != '.') {
					empty = 0;
				}

				if (input_board[i][j] == 'K') {
					chess.white_king.row = i;
					chess.white_king.col = j;
					chess.white_king.name = 'K';
				} else if (input_board[i][j] == 'k') {
					chess.black_king.row = i;
					chess.black_king.col = j;
					chess.black_king.name = 'k';
				}
			}
		}

		if (empty) break;
		game_count++;

		int white_in_check = is_in_check(chess, 1);
		int black_in_check = is_in_check(chess, 0);

		printf("Game #%d: ", game_count);
		if (white_in_check) {
			printf("white king is in check.\n");
		} else if (black_in_check) {
			printf("black king is in check.\n");
		} else {
			printf("no king is in check.\n");
		}
	}

	return 0;
}

File: pc01/pc01_07/python/pc01_07.py
"""
check the check
pc110107
uva10196
"""

from typing import List, Tuple, Set
import sys

SIZE = 8  # Board dimension (8x8)

# Movement directions
BLACK_PAWN_DIRS = [(-1, -1), (-1, 1)] + [(0, 0)] * 6
WHITE_PAWN_DIRS = [(1, -1), (1, 1)] + [(0, 0)] * 6
KNIGHT_DIRS = [(-2, -1), (-2, 1), (2, -1), (2, 1),
               (-1, -2), (-1, 2), (1, -2), (1, 2)]
ROOK_DIRS = [(-1, 0), (1, 0), (0, -1), (0, 1)] + [(0, 0)] * 4
BISHOP_DIRS = [(-1, -1), (-1, 1), (1, -1), (1, 1)] + [(0, 0)] * 4
QUEEN_DIRS = [(-1, -1), (-1, 0), (-1, 1),
              (0, -1), (0, 1),
              (1, -1), (1, 0), (1, 1)]

class ChessBoard:
    def __init__(self):
        self.board = [['.' for _ in range(SIZE)] for _ in range(SIZE)]
        self.white_king = (-1, -1)
        self.black_king = (-1, -1)
    
    def read_board(self) -> bool:
        is_empty = True
        for i in range(SIZE):
            row = input().strip()
            for j, piece in enumerate(row):
                self.board[i][j] = piece
                if piece != '.':
                    is_empty = False
                if piece == 'K':
                    self.white_king = (i, j)
                elif piece == 'k':
                    self.black_king = (i, j)
        return is_empty
    
    def is_valid_position(self, row: int, col: int) -> bool:
        return 0 <= row < SIZE and 0 <= col < SIZE
    
    def is_attacked_by_piece(self, king_pos: Tuple[int, int], piece_char: str, 
                           directions: List[Tuple[int, int]], repeating: bool) -> bool:
        king_row, king_col = king_pos
        
        for row_dir, col_dir in directions:
            if (row_dir, col_dir) == (0, 0):
                continue
                
            row, col = king_row, king_col
            
            if repeating:
                row += row_dir
                col += col_dir
                while self.is_valid_position(row, col):
                    if self.board[row][col] != '.':
                        if self.board[row][col] == piece_char:
                            return True
                        break
                    row += row_dir
                    col += col_dir
            else:
                row += row_dir
                col += col_dir
                if self.is_valid_position(row, col) and self.board[row][col] == piece_char:
                    return True
        
        return False
    
    def is_in_check(self, is_white: bool) -> bool:
        king_pos = self.white_king if is_white else self.black_king
        pieces = ['p' if is_white else 'P',  # Pawn
                 'n' if is_white else 'N',  # Knight
                 'r' if is_white else 'R',  # Rook
                 'b' if is_white else 'B',  # Bishop
                 'q' if is_white else 'Q',  # Queen
                 'k' if is_white else 'K']  # King
        
        # Check for pawn attacks
        pawn_dirs = BLACK_PAWN_DIRS if is_white else WHITE_PAWN_DIRS
        if self.is_attacked_by_piece(king_pos, pieces[0], pawn_dirs, False):
            return True
        
        # Check for knight attacks
        if self.is_attacked_by_piece(king_pos, pieces[1], KNIGHT_DIRS, False):
            return True
        
        # Check for rook attacks
        if self.is_attacked_by_piece(king_pos, pieces[2], ROOK_DIRS, True):
            return True
        
        # Check for bishop attacks
        if self.is_attacked_by_piece(king_pos, pieces[3], BISHOP_DIRS, True):
            return True
        
        # Check for queen attacks
        if self.is_attacked_by_piece(king_pos, pieces[4], QUEEN_DIRS, True):
            return True
        
        return False

def main():
    game_count = 0
    
    while True:
        chess = ChessBoard()
        if chess.read_board():
            break
        
        game_count += 1
        white_in_check = chess.is_in_check(True)
        black_in_check = chess.is_in_check(False)
        
        print(f"Game #{game_count}:", end=" ")
        if white_in_check:
            print("white king is in check.")
        elif black_in_check:
            print("black king is in check.")
        else:
            print("no king is in check.")

if __name__ == '__main__':
    main()